modules/sources/transformers/swap-v2-transformer.ts

```diff
@@ -14,10 +14,40 @@ export function swapV2Transformer(swap: BalancerSwapFragment, chain: Chain): Swa
     // Avoiding scientific notation
     const feeFloat = parseFloat(swap.tokenAmountIn) * parseFloat(swap.poolId.swapFee ?? 0);
     const fee = feeFloat < 1e6 ? feeFloat.toFixed(18).replace(/0+$/, '').replace(/\.$/, '') : String(feeFloat);
-    const feeFloatUSD = parseFloat(swap.valueUSD) * parseFloat(swap.poolId.swapFee ?? 0);
-    const feeUSD =
+    let feeFloatUSD = parseFloat(swap.valueUSD) * parseFloat(swap.poolId.swapFee ?? 0);
+    let feeUSD =
         feeFloatUSD < 1e6 ? feeFloatUSD.toFixed(18).replace(/0+$/, '').replace(/\.$/, '') : String(feeFloatUSD);
 
+    // FX pools have a different fee calculation
+    // Replica of the subgraph logic:
+    // https://github.com/balancer/balancer-subgraph-v2/blob/60453224453bd07a0a3a22a8ad6cc26e65fd809f/src/mappings/vault.ts#L551-L564
+    if (swap.poolId.poolType === 'FX') {
+        const USDC_ADDRESS = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';
+        const tokenOutAddress = swap.tokenOut;
+        const tokenInAddress = swap.tokenIn;
+        let isTokenInBase = tokenOutAddress == USDC_ADDRESS;
+        let baseTokenAddress = isTokenInBase ? tokenInAddress : tokenOutAddress;
+        let quoteTokenAddress = isTokenInBase ? tokenOutAddress : tokenInAddress;
+        let baseToken = swap.poolId.tokens?.find(({ token }) => token.address == baseTokenAddress);
+        let quoteToken = swap.poolId.tokens?.find(({ token }) => token.address == quoteTokenAddress);
+        let baseRate = baseToken != null ? baseToken.token.latestFXPrice : null;
+        let quoteRate = quoteToken != null ? quoteToken.token.latestFXPrice : null;
+
+        if (baseRate && quoteRate) {
+            if (isTokenInBase) {
+                feeFloatUSD +=
+                    parseFloat(swap.tokenAmountIn) * parseFloat(baseRate) -
+                    parseFloat(swap.tokenAmountOut) * parseFloat(quoteRate);
+            } else {
+                feeFloatUSD +=
+                    parseFloat(swap.tokenAmountIn) * parseFloat(quoteRate) -
+                    parseFloat(swap.tokenAmountOut) * parseFloat(baseRate);
+            }
+        }
+
+        feeUSD = String(feeFloatUSD);
```

Comment:
Think we talked about the option of a possible negative APR. Would that mean that his fee might be negative? What are the downstream effects if so?

Reply:
is there upside / downside? i assume it's just a property of FX pools. unless you mean if some other parts of the app will need to be updated. that i'm not sure. Is there anything that depends on the apr?